src/services/scope-manager.js

```diff
@@ -0,0 +1,76 @@
+import context from '../context';
+
+// 5 minutes exipration cache
+const SCOPE_CACHE_EXPIRATION_DELTA = 300;
+
+class ScopeManager {
+  constructor({ configStore, forestServerRequester, moment } = context.inject()) {
+    this.configStore = configStore;
+    this.forestServerRequester = forestServerRequester;
+    this.moment = moment;
+    this.scopesCache = {};
+  }
+
+  async getScopeForUser(user, collectionName) {
+    if (!user.renderingId) throw new Error('Missing required renderingId');
+    if (!collectionName) throw new Error('Missing required collectionName');
+
+    const collectionScope = await this._getScopeCollectionScope(user.renderingId, collectionName);
+
+    return ScopeManager._formatDynamicValues(user.id, collectionScope);
+  }
+
+  static _formatDynamicValues(userId, collectionScope) {
+    if (!collectionScope?.scope?.filter) return null;
+
+    collectionScope.scope.filter.conditions.forEach((condition) => {
+      if (condition.value
+        && `${condition.value}`.startsWith('$')
+        && collectionScope.scope.dynamicScopesValues.users[userId]) {
+        condition.value = collectionScope
+          .scope
+          .dynamicScopesValues
+          .users[userId][condition.value];
+      }
+    });
+
+    return collectionScope.scope.filter;
+  }
+
+  async _getScopeCollectionScope(renderingId, collectionName) {
+    // if scopes have never been fetched before, wait for those to be retrieved
+    if (!this.scopesCache[renderingId]) {
+      await this._refreshScopesCache(renderingId);
+    } else if (this._hasCacheExpired(renderingId)) {
+    // if cache expired fetch scopes but don't wait for it
+      this._refreshScopesCache(renderingId);
+    }
+
+    return this.scopesCache[renderingId].scopes[collectionName];
+  }
+
+  async _fetchScopes(renderingId) {
+    const { envSecret } = this.configStore.lianaOptions;
+    const queryParams = { renderingId };
+
+    return this.forestServerRequester.perform('/liana/scopes', envSecret, queryParams);
+  }
+
+  async _refreshScopesCache(renderingId) {
+    const scopes = await this._fetchScopes(renderingId);
```

Comment:
there is a race condition here.

if the cache is outdated, and the method is called many times in a short burst, the query will be triggered many times.
You should update the time in the cache before, then make the query